File: ml-benchmark/synduce/constraints/bst/minimax_separate.f
Inductive Tree = leaf Int | node {Int, Tree, Tree};

min = \a: Int. \b: Int. if < a b then a else b;
max = \a: Int. \b: Int. if < a b then b else a;

tmin = fix (
  \f: Tree -> Int. \t: Tree.
  match t with
    leaf w -> w
  | node {w, l, r} -> min w (min (f l) (f r))
  end
);

tmax = fix (
  \f: Tree -> Int. \t: Tree.
  match t with
    leaf w -> w
  | node {w, l, r} -> max w (max (f l) (f r))
  end
);

is_bst = fix (
  \f: Tree -> Bool. \t: Tree.
  match t with
    leaf w -> true
  | node {w, l, r} -> and (and (>= w (tmax l)) (<= w (tmin r))) (and (f l) (f r))
  end
);


spec = fix (
  \f: Tree -> {Int, Int}. \t: Tree.
  match t with
    leaf x -> {x, x}
  | node {a, l, r} ->
    let lres = f l in let rres = f r in
      {max a (max lres.1 rres.1), min a (min lres.2 rres.2)}
  end
);

target = fix (
  \f: Tree -> Compress Tree. \t: Tree.
  match t with
    leaf x -> leaf x
  | node {a, l, r} ->
    /*These invocations are provided in Synduce's template*/
    let lmin = tmin l in let lmax = tmax l in
    let rmin = tmin r in let rmax = tmax r in
      node {a, l, r} /*Avoid recursions*/
  end
);

main = \t: Tree. if is_bst t then spec (target t) else {0, 0};

File: ml-benchmark/synduce/constraints/bst/minimax_separate.ml
type tree = Leaf of int | Node of int * tree * tree

let min a b = if a < b then a else b
let max a b = if a < b then b else a

let rec tmin t =
  match t with
  | Leaf w -> w
  | Node (w, l, r) -> min w (min (tmin l) (tmin r))

let rec tmax t =
  match t with
  | Leaf w -> w
  | Node (w, l, r) -> max w (max (tmax l) (tmax r))

let rec is_bst t =
  match t with
  | Leaf w -> true
  | Node (w, l, r) ->
      ((w >= tmax l) && (w <= tmin r)) && (is_bst l && is_bst r)

let rec spec t =
  match t with
  | Leaf x -> (x, x)
  | Node (a, l, r) ->
      let lres = spec l in
      let rres = spec r in
      (max a (max (fst lres) (fst rres)), min a (min (snd lres) (snd rres)))

let target t =
  match t with
  | Leaf x -> Leaf x
  | Node (a, l, r) ->
      let lmin = tmin l in
      let lmax = tmax l in
      let rmin = tmin r in
      let rmax = tmax r in
      Node (a, l, r)

let program t = if is_bst t then spec (target t) else (0, 0)